### PR TITLE
error handling: send BAD_CREDENTIALS when a login required/logged out error occurs

### PR DESCRIPTION
**Pull request by sumnerevans**
Signed-off-by: Sumner Evans <sumner@beeper.com>

### CODE REVIEW

File: mauigpapi/http/base.py [hunk 1]
```diff
@@ -239,22 +241,25 @@ async def _handle_response(self, resp: ClientResponse) -> JSON:
         if body.get("status", "fail") == "ok":
             return body
         else:
-            await self._raise_response_error(resp)
+            err = await self._get_response_error(resp)
+            if self.on_response_error:
+                self.on_response_error(err)
```
Comment (tulir) [inline mauigpapi/http/base.py:246]:
```suggestion
                await self.on_response_error(err)
```

File: mautrix_instagram/user.py [hunk 2]
```diff
@@ -376,6 +380,18 @@ async def on_proxy_update(self, evt: ProxyUpdate | None = None) -> None:
...
+        if isinstance(err, (IGLoginRequiredError, IGUserHasLoggedOutError)):
+            self.log.warning("User got logged out")
+            self._track_metric(METRIC_CONNECTED, False)
+            self._is_connected = False
+            await self.push_bridge_state(
+                BridgeStateEvent.BAD_CREDENTIALS,
+                error="logged-out",
+                info={"cnd_action": "reauth"},
+            )
+            return
```
Comment (tulir) [inline mautrix_instagram/user.py:393]:
```suggestion
            self.log.warning(f"Noticed logout in API error response: {err}")
            await self.logout(error=err)
```

File: mautrix_instagram/user.py [hunk 3]
```diff
@@ -376,6 +380,18 @@ async def on_proxy_update(self, evt: ProxyUpdate | None = None) -> None:
         if self.command_status:
             self.command_status["api"].setup_http(self.command_status["state"].cookies.jar)
 
+    async def on_response_error(self, err: IGResponseError) -> None:
+        if isinstance(err, (IGLoginRequiredError, IGUserHasLoggedOutError)):
```
Comment (tulir) [inline mautrix_instagram/user.py:384]:
```suggestion
        if isinstance(err, IGNotLoggedInError):
```

**Comment by tulir** [comment]
Just realized that this will also conflict with the normal auth checks (like checking if the user is logged in) 🤔

File: mautrix_instagram/user.py [hunk 4]
```diff
@@ -376,6 +378,11 @@ async def on_proxy_update(self, evt: ProxyUpdate | None = None) -> None:
         if self.command_status:
             self.command_status["api"].setup_http(self.command_status["state"].cookies.jar)
 
+    async def on_response_error(self, err: IGResponseError) -> None:
+        if isinstance(err, IGNotLoggedInError):
```
Comment (Fizzadar) [inline mautrix_instagram/user.py:382]:
```suggestion
        if isinstance(err, IGNotLoggedInError) and (await self.is_logged_in()):
```

@tulir this should resolve the issue with the login checks I think?

Comment (tulir) [reply mautrix_instagram/user.py:382]: It doesn't resolve calls like the one in `fetch_user_and_reconnect`, `_sync`, `_handle_backfill_requests_loop` and `run_with_sync_lock`, but maybe we could just remove those (and assume that the response error handler always handles it)

**Pull request merged**